docs/core-concepts/params.md

```diff
@@ -0,0 +1,168 @@
+| -------- | ------- | --------------------------------------------------------------------------------------------- |
+| `String` | ✓       | Any string value                                                                              |
+| `Number` |         | Any value that can be parsed with the `Number` constructor                                    |
+| `Boolean`|         | The literal values `"true"` or `"false"`                                                      |
+| `Date`   |         | Any value that can be parsed with the `Date` constructor. Uses `toISOString` when serializing |
+| `RegExp` |         | A literal regex expression                                                                    |
+| `JSON`   |         | Any value that can be parsed with the `JSON.parse` method                                     |
+
+## Custom Param Types
+You can define your own param types using the `createParam` utility. This serves three purposes:
+1. Define the type of the param.
+2. Define how the param should be serialized and deserialized.
+3. Determine if a value is valid.
+
+Using custom param types you can define more complex validation rules. For example, this month param must be a valid month name. If it is not, the route will not match.
+
+```ts {3-11,17}
+import { createParam, createRoute, path } from '@kitbag/router'
+
+const months = ['january', 'february', 'march', 'april', 'may', 'june', 'july', 'august', 'september', 'october', 'november', 'december'] as const
+
+const monthParam = createParam((value, { invalid }) => {
+  if (months.includes(value as any)) {
+    return value
+  }
+
+  invalid(`Invalid month: ${value}`)
+})
+
+const events = createRoute({
+  name: 'events',
+  path: path('/events/[year]/[?month]', {
+    year: Number,
+    month: monthParam,
+  }),
+})
+```
+
+By default custom params will use `toString` when serializing. But you can use a Get/Set param to define how the param should be serialized and deserialized. For example we can make the month param capitalized when parsing and lowercase when serializing.
+
+```ts
+import { createParam } from '@kitbag/router'
+
+const months = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'] as const
+
+const monthParam = createParam({
+  get: (value, { invalid }) => {
+    const month = value.charAt(0).toUpperCase() + value.slice(1)
+
+    if (months.includes(month as any)) {
+      return month
+    }
+
+    invalid(`Invalid month: ${month}`)
+  },
+  set: (value) => {
+    return value.toLowerCase()
+  },
+})
+```
+
+## Default Values
+You can define a default value for a param type when creating it. This value will be used when the param is optional and not provided in the url. An optional param of type `String` will have the type `string | undefined` when being accessed. But when a default value is provided it will have the type `string`, even if the value is missing from the url.
```

Comment:
```suggestion
You can define a default value for a param type when defining  it. This default value will be used when the param is optional and not provided in the url. An optional param of type `String` will have the type `string | undefined` when being accessed. But when a default value is provided it will have the type `string`, even if the value is missing from the url.
```

Reply:
This is splitting hairs a bit, I'd take your version but there's an extra space in it lol